In [26]:
import sys
import os
import re
import string
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import psycopg2
from sqlalchemy import create_engine
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import ElasticNet
from sklearn.metrics import mean_squared_error
import warnings
warnings.filterwarnings('ignore')

# import tensorflow as tf
# from tensorflow.keras.models import Sequential
# from tensorflow.keras.layers import Dense, LSTM, Dropout, Embedding
# from tensorflow.keras.preprocessing import sequence
# from imblearn.over_sampling import RandomOverSampler # TIP: If error occurs, restart kernal
# from keras.preprocessing.text import Tokenizer
# from tensorflow.keras.layers import TextVectorization
# from keras.preprocessing.text import text_to_word_sequence

In [2]:
# engine = create_engine('postgresql://postgres:postgres@localhost/proj_4')
engine = psycopg2.connect('postgresql://postgres:postgres@localhost/proj_4')

# Explatory Data Analysis

In [3]:
cars_df = pd.read_sql('select * from used_cars', engine)
cars_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80170 entries, 0 to 80169
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         80170 non-null  float64
 1   year          80170 non-null  int64  
 2   manufacturer  80170 non-null  object 
 3   condition     80170 non-null  object 
 4   cylinders     80170 non-null  object 
 5   fuel          80170 non-null  object 
 6   odometer      80170 non-null  float64
 7   title_status  80170 non-null  object 
 8   transmission  80170 non-null  object 
 9   drive         80170 non-null  object 
 10  size          80170 non-null  object 
 11  type          80170 non-null  object 
 12  paint_color   80170 non-null  object 
 13  state         80170 non-null  object 
 14  posting_date  80170 non-null  object 
dtypes: float64(2), int64(1), object(12)
memory usage: 9.2+ MB


In [4]:
cars_df.head()

price  year manufacturer  condition    cylinders    fuel  odometer  \
0  15000.0  2013         ford  excellent  6 cylinders     gas  128000.0   
1  19900.0  2004         ford       good  8 cylinders  diesel   88000.0   
2  14000.0  2012        honda  excellent  6 cylinders     gas   95000.0   
3  22500.0  2001         ford       good  8 cylinders  diesel  144700.0   
4  15000.0  2017        dodge  excellent  8 cylinders     gas   90000.0   

  title_status transmission drive       size      type paint_color state  \
0        clean    automatic   rwd  full-size     truck       black    al   
1        clean    automatic   4wd  full-size    pickup        blue    al   
2        clean    automatic   fwd  full-size  mini-van      silver    al   
3        clean       manual   rwd  full-size     truck       white    al   
4      rebuilt    automatic   rwd   mid-size     sedan        grey    al   

  posting_date  
0   2021-05-03  
1   2021-04-29  
2   2021-04-27  
3   2021-04-26  
4   2021-04-24

In [5]:
# Drop posting date col
df = cars_df.drop(columns=['posting_date'])
df.head()

price  year manufacturer  condition    cylinders    fuel  odometer  \
0  15000.0  2013         ford  excellent  6 cylinders     gas  128000.0   
1  19900.0  2004         ford       good  8 cylinders  diesel   88000.0   
2  14000.0  2012        honda  excellent  6 cylinders     gas   95000.0   
3  22500.0  2001         ford       good  8 cylinders  diesel  144700.0   
4  15000.0  2017        dodge  excellent  8 cylinders     gas   90000.0   

  title_status transmission drive       size      type paint_color state  
0        clean    automatic   rwd  full-size     truck       black    al  
1        clean    automatic   4wd  full-size    pickup        blue    al  
2        clean    automatic   fwd  full-size  mini-van      silver    al  
3        clean       manual   rwd  full-size     truck       white    al  
4      rebuilt    automatic   rwd   mid-size     sedan        grey    al

In [6]:
# check number of unique values in each column
cats = {}
for i in df.columns:
    cats[i] = df[i].nunique()

print(cats)

{'price': 4257, 'year': 104, 'manufacturer': 41, 'condition': 6, 'cylinders': 8, 'fuel': 5, 'odometer': 28224, 'title_status': 6, 'transmission': 3, 'drive': 3, 'size': 4, 'type': 13, 'paint_color': 12, 'state': 51}


### Check Normality

In [9]:
# Select columns to check for normality (continuous numeric variables) - price & odometer
numeric_columns = ['price', 'odometer']

# Loop through the selected columns and perform normality tests
for col in numeric_columns:
    stat, p = stats.normaltest(df[col])

    print(f"Column: {col}")
    print("Statistic:", stat)
    print("p-value:", p)

    if p < 0.05:
        print("The data does not appear to be normally distributed.\n")
    else:
        print("The data appears to be normally distributed.\n")

Column: price
Statistic: 419772.357151128
p-value: 0.0
The data does not appear to be normally distributed.

Column: odometer
Statistic: 218332.90779230325
p-value: 0.0
The data does not appear to be normally distributed.



# Preprocessing the Data

## Encoding

In [7]:
# One-hot encoding for categorical variables
dummies_df = pd.get_dummies(df, columns=['manufacturer', 'condition', 'cylinders', 'fuel', 'title_status', 'transmission', 'drive', 'size', 'type', 'paint_color', 'state'])
dummies_df.head()

price  year  odometer  manufacturer_acura  manufacturer_alfa-romeo  \
0  15000.0  2013  128000.0                   0                        0   
1  19900.0  2004   88000.0                   0                        0   
2  14000.0  2012   95000.0                   0                        0   
3  22500.0  2001  144700.0                   0                        0   
4  15000.0  2017   90000.0                   0                        0   

   manufacturer_aston-martin  manufacturer_audi  manufacturer_bmw  \
0                          0                  0                 0   
1                          0                  0                 0   
2                          0                  0                 0   
3                          0                  0                 0   
4                          0                  0                 0   

   manufacturer_buick  manufacturer_cadillac  ...  state_sd  state_tn  \
0                   0                      0  ...         0         0   
1                   0                      0  ...         0         0   
2                   0                      0  ...         0         0   
3                   0                      0  ...         0         0   
4                   0                      0  ...         0         0   

   state_tx  state_ut  state_va  state_vt  state_wa  state_wi  state_wv  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   state_wy  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 155 columns]

## Feature Scaling

In [8]:
# Scale the numeric data
# No need to scale price since it is our target variable
# No need to scale year since it is a discrete numeric variable?
scaled_data = StandardScaler().fit_transform(df[['odometer']])
scaled_data_df = pd.DataFrame(scaled_data, columns=['odometer'])
scaled_data_df.head()

dummies_df['odometer'] = scaled_data_df['odometer']

dummies_df.head()

price  year  odometer  manufacturer_acura  manufacturer_alfa-romeo  \
0  15000.0  2013  0.012768                   0                        0   
1  19900.0  2004 -0.147564                   0                        0   
2  14000.0  2012 -0.119506                   0                        0   
3  22500.0  2001  0.079706                   0                        0   
4  15000.0  2017 -0.139547                   0                        0   

   manufacturer_aston-martin  manufacturer_audi  manufacturer_bmw  \
0                          0                  0                 0   
1                          0                  0                 0   
2                          0                  0                 0   
3                          0                  0                 0   
4                          0                  0                 0   

   manufacturer_buick  manufacturer_cadillac  ...  state_sd  state_tn  \
0                   0                      0  ...         0         0   
1                   0                      0  ...         0         0   
2                   0                      0  ...         0         0   
3                   0                      0  ...         0         0   
4                   0                      0  ...         0         0   

   state_tx  state_ut  state_va  state_vt  state_wa  state_wi  state_wv  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   state_wy  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 155 columns]

### Checking normality again

In [10]:

# Select columns to check for normality (continuous numeric variables) - price & odometer
numeric_columns = ['price', 'odometer']

# Loop through the selected columns and perform normality tests
for col in numeric_columns:
    stat, p = stats.normaltest(dummies_df[col])

    print(f"Column: {col}")
    print("Statistic:", stat)
    print("p-value:", p)

    if p < 0.05:
        print("The data does not appear to be normally distributed.\n")
    else:
        print("The data appears to be normally distributed.\n")

Column: price
Statistic: 419772.357151128
p-value: 0.0
The data does not appear to be normally distributed.

Column: odometer
Statistic: 218332.90779230325
p-value: 0.0
The data does not appear to be normally distributed.



In [16]:
# Since the price and odometer are not normally distributed, we will find Spearman Correlation Coefficients
stats.spearmanr(dummies_df['price'], dummies_df['odometer'])

SpearmanrResult(correlation=-0.3952764744998881, pvalue=0.0)

0.395 is considered low correlation

# Split Train Test

In [21]:
# Separate the features (X) from the Target(y)
y = dummies_df["price"]
X = dummies_df.drop(columns="price")

# Split data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1)
X_train.shape

(60127, 154)

# Elastic Net Regression Model

In [36]:
# Create an Elastic Net model
elastic_net = ElasticNet(alpha=1.0, l1_ratio=0.5)  # Adjust alpha and l1_ratio as needed

# Train the model on the scaled training data
elastic_net.fit(X_train, y_train)

# Make predictions on the scaled test data
y_pred_elastic = elastic_net.predict(X_test)

# Calculate the Mean Squared Error (MSE) for evaluation
mse_elastic = mean_squared_error(y_test, y_pred_elastic)
print(f"Mean Squared Error: {mse_elastic:.2f}")
print(f"Root Mean Squared Error: {np.sqrt(mse_elastic):.2f}")

# Calculate the R-squared value using the score method
r_squared_elastic = elastic_net.score(X_test, y_test)
print(f"R-squared value: {r_squared_elastic:.2f}")


Mean Squared Error: 39727037425.91
Root Mean Squared Error: 199316.43
R-squared value: -0.71


## Hyper Parameter Tuning

`alpha` controls the overall strength of regularization <br>
`l1_ratio` controls the balance between L1 and L2 regularization

In [27]:
# Define the hyperparameter grid
param_grid = {
    'alpha': [0.01, 0.1, 1.0, 10.0],  # Values for alpha
    'l1_ratio': [0.1, 0.3, 0.5, 0.7, 0.9]  # Values for l1_ratio
}

# Create an Elastic Net model
elastic_net = ElasticNet()

# Perform grid search with cross-validation
grid_search = GridSearchCV(elastic_net, param_grid, cv=5)
grid_search.fit(X, y)

# Get the best hyperparameters
best_alpha = grid_search.best_params_['alpha']
best_l1_ratio = grid_search.best_params_['l1_ratio']

print(f"Best alpha: {best_alpha}")
print(f"Best l1_ratio: {best_l1_ratio}")

Best alpha: 10.0
Best l1_ratio: 0.1


In [37]:
# Create an Elastic Net model
elastic_net = ElasticNet(alpha=10.0, l1_ratio=0.1)  # Adjust alpha and l1_ratio as needed

# Train the model on the scaled training data
elastic_net.fit(X_train, y_train)

# Make predictions on the scaled test data
y_pred_elastic = elastic_net.predict(X_test)

# Calculate the Mean Squared Error (MSE) for evaluation
mse_elastic = mean_squared_error(y_test, y_pred_elastic)
print(f"Mean Squared Error: {mse_elastic:.2f}")
print(f"Root Mean Squared Error: {np.sqrt(mse_elastic):.2f}")

# Calculate the R-squared value using the score method
r_squared_elastic = elastic_net.score(X_test, y_test)
print(f"R-squared value: {r_squared_elastic:.2f}")

Mean Squared Error: 31294213348.00
Root Mean Squared Error: 176901.71
R-squared value: -0.35


Increasing the value of alpha and moving l1_ratio towards 0.1 will lead to stronger overall regularization and a bias towards Ridge-like regularization in the Elastic Net model. This did end up reducing the mean squared error, which means the tuned model is better. However, the model's predictions are, on average, $176,901.71 far from the actual values. The R-squared value is negative before tuning and after tuning, which suggests that the model does not fit the Data.

# Random Forest Regression